src/py/flwr/server/superlink/fleet/vce/vce_api.py

```diff
@@ -42,21 +46,217 @@ def _register_nodes(
+            context = node_states[node_id].retrieve_context(run_id=task_ins.run_id)
+
+            # Convert TaskIns to Message
+            message = message_from_taskins(task_ins)
+            # Replace node ID with data partition ID
+            message.metadata.partition_id = nodes_mapping[node_id]
+
+            # Let backend process message
+            out_mssg, updated_context = await backend.process_message(
+                app, message, context
+            )
+
+            # Update Context
+            node_states[node_id].update_context(
+                task_ins.run_id, context=updated_context
+            )
+
+            # Convert to TaskRes
+            task_res = message_to_taskres(out_mssg)
+            # Store TaskRes in state
+            state.store_task_res(task_res)
+
+        except asyncio.CancelledError as e:
+            log(DEBUG, "Async worker: %s", e)
+            break
+
+        except LoadClientAppError as app_ex:
+            log(ERROR, "Async worker: %s", app_ex)
+            log(ERROR, traceback.format_exc())
+            raise
+
+        except Exception as ex:  # pylint: disable=broad-exception-caught
+            log(ERROR, ex)
+            log(ERROR, traceback.format_exc())
+            break
+
+
+async def add_taskins_to_queue(
+    queue: "asyncio.Queue[TaskIns]",
+    state_factory: StateFactory,
+    nodes_mapping: NodeToPartitionMapping,
+    backend: Backend,
+    consumers: List["asyncio.Task[None]"],
+    f_stop: asyncio.Event,
+) -> None:
+    """Retrieve TaskIns and add it to the queue."""
+    state = state_factory.state()
+    num_initial_consumers = len(consumers)
+    while not f_stop.is_set():
+        for node_id in nodes_mapping.keys():
+            task_ins = state.get_task_ins(node_id=node_id, limit=1)
+            if task_ins:
+                await queue.put(task_ins[0])
+
+        # Count consumers that are running
+        num_active = sum(not (cc.done()) for cc in consumers)
+
+        # Alert if number of consumers decreased by half
+        if num_active < num_initial_consumers // 2:
+            log(
+                WARN,
+                "Number of active workers has more than halved: (%i/%i active)",
+                num_active,
+                num_initial_consumers,
+            )
+
+        # Break if consumers died
+        if num_active == 0:
+            raise RuntimeError("All workers have died. Ending Simulation.")
+
+        # Log some stats
+        log(
+            DEBUG,
+            "Simulation Engine stats: "
+            "Active workers: (%i/%i) | %s (%i workers) | Tasks in queue: %i)",
+            num_active,
+            num_initial_consumers,
+            backend.__class__.__name__,
+            backend.num_workers,
+            queue.qsize(),
+        )
+        await asyncio.sleep(1.0)
+    log(DEBUG, "Async producer: Stopped pulling from StateFactory.")
+
+
+async def run(
+    app: Callable[[], ClientApp],
```

Comment:
Should we renamed this to sth like `app_fn` or `app_factory` for clarity?
```suggestion
    app_fn: Callable[[], ClientApp],
```